packages/@aws-cdk/aws-elasticsearch/lib/domain.ts

```diff
@@ -748,90 +820,494 @@ export class Domain extends cdk.Resource implements IDomain {
+
+
+  private readonly domain: CfnDomain;
+  private readonly slowSearchLogGroup?: logs.ILogGroup;
+  private readonly slowIndexLogGroup?: logs.ILogGroup;
+  private readonly appLogGroup?: logs.ILogGroup;
+
+  constructor(scope: cdk.Construct, id: string, props: DomainProps) {
+    super(scope, id, {
+      physicalName: props.domainName,
+    });
+
+    // If VPC options are supplied ensure that the number of subnets matches the number AZ
+    if (props.vpcOptions?.subnets.map((subnet) => subnet.availabilityZone).length != props?.clusterConfig.availabilityZoneCount) {
+      throw new Error('When providing vpc options you need to provide a subnet for each AZ you are using');
+    };
+
+    const masterInstanceType = props.clusterConfig.masterNodeInstanceType.toLowerCase();
+    const dataInstanceType = props.clusterConfig.dataNodeInstanceType.toLowerCase();
+
+    if ([masterInstanceType, dataInstanceType].some(instanceType => !instanceType.endsWith('.elasticsearch'))) {
+      throw new Error('Master and data node instance types must end with ".elasticsearch".');
+    }
+
+    const elasticsearchVersion = props.elasticsearchVersion ?? ElasticsearchVersion.ES_VERSION_7_4;
+    const versionNumber = parseFloat(elasticsearchVersion.toString());
+    const encryptionAtRestEnabled = props.encryptionAtRestOptions?.enabled ?? (props.encryptionAtRestOptions?.kmsKey != null);
+    const ebsEnabled = props.ebsOptions != null;
+
+    const isInstanceType = function(instanceType: string) : Boolean {
+      return masterInstanceType.startsWith(instanceType) || dataInstanceType.startsWith(instanceType);
+    };
+
+    const isSomeInstanceType = function(...instanceTypes: string[]) : Boolean {
+      return instanceTypes.some(isInstanceType);
+    };
+
+    const isEveryInstanceType = function(...instanceTypes: string[]) : Boolean {
+      return instanceTypes.every(isInstanceType);
+    };
+
+    // Validate feature support for the given Elasticsearch version, per
+    // https://docs.aws.amazon.com/elasticsearch-service/latest/developerguide/aes-features-by-version.html
+    if (versionNumber < 5.1) {
+      if (
+        props.logPublishingOptions?.slowIndexLogEnabled
+        || props.logPublishingOptions?.appLogEnabled
+        || props.logPublishingOptions?.slowSearchLogEnabled
+      ) {
+        throw new Error('Error and slow logs publishing requires Elasticsearch version 5.1 or later.');
+      }
+      if (props.encryptionAtRestOptions?.enabled) {
+        throw new Error('Encryption of data at rest requires Elasticsearch version 5.1 or later.');
+      }
+      if (props.cognitoOptions != null) {
+        throw new Error('Cognito authentication for Kibana requires Elasticsearch version 5.1 or later.');
+      }
+      if (isSomeInstanceType('c5', 'i3', 'm5', 'r5')) {
+        throw new Error('C5, I3, M5, and R5 instance types require Elasticsearch version 5.1 or later.');
+      }
+    } else if (versionNumber < 5.3) {
```

Comment:
This is not right. See https://docs.aws.amazon.com/elasticsearch-service/latest/developerguide/es-managedomains-snapshots.html

Below 5.3 daily snapshots are taken and the user specifies `automatedSnapshotStartHour`. With 5.3 this setting is ignored and hourly snapshots are taken.

Reply:
I don't think these can be `else if` since the changes stack. Currently if I have a version number < 5.1 then this would not be checked.

Reply:
🤦 good catch!

Re: the first comment, you're right. I was referencing [this doc](https://docs.aws.amazon.com/elasticsearch-service/latest/developerguide/aes-features-by-version.html) and it threw me off.

Reply:
Yeah I can see that :).